nautobot/ipam/forms.py

```diff
@@ -713,16 +713,29 @@ def clean(self):
         # Cannot select both a device interface and a VM interface
         if self.cleaned_data.get("interface") and self.cleaned_data.get("vminterface"):
             raise forms.ValidationError("Cannot select both a device interface and a virtual machine interface")
+
+        # Stash a copy of `assigned_object` before we replace it, so we can use it in `save()`.
+        self.instance._original_assigned_object = self.instance.assigned_object
         self.instance.assigned_object = self.cleaned_data.get("interface") or self.cleaned_data.get("vminterface")
 
         # Primary IP assignment is only available if an interface has been assigned.
         interface = self.cleaned_data.get("interface") or self.cleaned_data.get("vminterface")
-        if self.cleaned_data.get("primary_for_parent") and not interface:
+        primary_for_parent = self.cleaned_data.get("primary_for_parent")
+        if primary_for_parent and not interface:
             self.add_error(
                 "primary_for_parent",
                 "Only IP addresses assigned to an interface can be designated as primary IPs.",
             )
 
+        # If `primary_for_parent` is unset, clear the `primary_ip{version}` for the
+        # Device/VirtualMachine. It will not be saved until after `IPAddress.clean()` succeeds which
+        # also checks for the `_primary_ip_unset_by_form` value.
+        if not primary_for_parent and hasattr(self.instance._original_assigned_object, "parent"):
+            ip_version = self.instance.address.version
+            parent = self.instance._original_assigned_object.parent
+            setattr(parent, f"primary_ip{ip_version}", None)  # e.g. `primary_ip4` or `primary_ip6`
```

Comment:
To me it seems like it'd be cleaner if these three lines were moved to the `if primary_ip_unset_by_form` block in `save()`  - that way `clean()` doesn't directly modify the device, only `save()` does.

Reply:
Ooh, that's ugly indeed. I don't like `clean()` causing changes to the data, but I guess if `save()` never gets called, the changes are purely in-memory and don't actually affect the DB.

Reply:
Hang on, I'm still missing something here. The check in `IPAddress.clean()` is gated now by `_primary_ip_unset_by_form`. So why isn't just setting that flag here sufficient - why do we also need to clear out the `parent.primary_ip` now instead of deferring that action to the form `save()`?

Reply:
> Ooh, that's ugly indeed. I don't like `clean()` causing changes to the data, but I guess if `save()` never gets called, the changes are purely in-memory and don't actually affect the DB.

Well `clean()` was already changing data, like  on line 719 (`self.instance.assigned_object = self.cleaned_data.get("interface") or self.cleaned_data.get("vminterface")`). And I don't see that as a problem as long as it's not saved until all validation passes.

Reply:
> Hang on, I'm still missing something here. The check in `IPAddress.clean()` is gated now by `_primary_ip_unset_by_form`. So why isn't just setting that flag here sufficient - why do we also need to clear out the `parent.primary_ip` now instead of deferring that action to the form `save()`?

Ahhhhhh yeah okay, right, I think your'e right. I was still a little stuck in the past. Let me revisit it.